book/src/internals/build-system.md

```diff
@@ -0,0 +1,93 @@
+
+## The problems (or "challenges" if you prefer corporate jargon 😉)
+
+We unfortunately cannot simply link your Rust code into a static library and link to it for the following reasons:
+
+### Static Initializers
+
+Qt code often contains initialization code that is called by a static variable that runs the initialization code in its constructor.
+
+However, when linking into a static library, and then linking into the main executable, the linker will discard everything from the library that isn't used by the main executable.
+
+There are two ways to solve this:
+- Export an object file and link that to the main binary. Object files are always included completely
+- Use the whole-archive linker flag which forces inclusion of every object within the static library.
+  - If we include the entire static lib generated by cargo, then we'll likely get duplicate symbols, as this really includes **everything** that your Rust code **may** need, even if you don't use it.
+  - This has caused some recent regressions with Rust 1.78+, where MSVC could no longer link CXX-Qt due to duplicate symbols
+  - The way to solve this is to only export the static initializers as a library and link that into CMake.
+
+### Header files
+
+We want to make the generated headers available, not just to CMake, but also within dependents in the cargo build chain (e.g. your crate will probably want to depend on the headers produced by cxx-qt-lib).
+
+For this we need to export them to a stable directory so that both CMake and Cargo can find them.
+
+### (Optional) Integration with CMake
+
+Somehow, all of this should be compatible with both CMake, and Cargo-only builds.
+
+## The plan (for now)
+
+After many rounds of refactoring this, we believe the best way to go is to not rely on cargos OUT_DIR exclusivly.
+Not being able to share artifacts between crates is very limiting.
+
+We want to use a similar approach to CXX, which stores its data within either:
+- Cargos `target/` directory under the `cxxbridge` subfolder.
+- A custom `scratch` directory
+
+See: [https://github.com/dtolnay/cxx/blob/afd4aa3f3d4e5d5e9a3a41d09df3408f5f86a469/gen/build/src/target.rs#L10](https://github.com/dtolnay/cxx/blob/afd4aa3f3d4e5d5e9a3a41d09df3408f5f86a469/gen/build/src/target.rs#L10) \
+and: [https://github.com/dtolnay/cxx/blob/afd4aa3f3d4e5d5e9a3a41d09df3408f5f86a469/gen/build/src/lib.rs#L178](https://github.com/dtolnay/cxx/blob/afd4aa3f3d4e5d5e9a3a41d09df3408f5f86a469/gen/build/src/lib.rs#L178).
+
+We will likely mirror this and use the `cxxqtbridge` naming.
+
+## Contents of `cxxqtbridge`
+
+### `crates` directory
+
+Inside cxxqtbridge, there should be a `crates` folder with one subfolder per crate.
+Each crates subfolder should contain the following:
+
+- `include/`
+  - `crate-name` - A folder for all headers that are exported by this crate
+  - `cxx-qt-lib -> ../../cxx-qt-lib/include` - Symbolic links for every dependency
```

Comment:
ah yes good point then we can avoid having multiple copies like at the moment, also note that we have the `rust/cxx.h` that needs to be injected somewhere?

Reply:
Rust, for the cxx.h we'll have to see where to put that, maybe we do duplicate that one...

Reply:
This is done now ? as we now know where it ends up?

Reply:
It is still duplicated into the include directory of every crate, but shouldn't bee too much of an issue.